Create development guide section (#7104)
kubernetes/ingress-nginx: docs/developer-guide/code-overview.md

# Ingress NGINX - Code Overview

This document provides an overview of Ingress NGINX code

## Core Golang code

This part of the code is responsible for the main logic of Ingress NGINX. It contains all the logics that parses Ingress Objects, annotations, watches Endpoints and turn them into usable nginx.conf configuration

```TODO
TODO: Add links for Ingress, Endpoints, annotations concepts
```

### Core Sync Logics:

Ingress-nginx has an internal model of the ingresses, secrets and endpoints in a given cluster. It maintains two copy of that (1) currently running configuration model and (2) the one generated in response to some changes in the cluster. 

The sync logic diffs the two models and if there's a change it tries to converge the running configuration to the new one. 

There are static and dynamic configuration changes. 

All endpoints and certificate changes are handled dynamically by posting the payload to an internal NGINX endpoint that is handled by Lua.

---

The following parts of the code can be found:

> ### Entrypoint

Reviewer: ```suggestion
### Entrypoint

```

Is the `main` package, responsible for starting ingress-nginx program.

It can be found in [cmd/nginx](https://github.com/kubernetes/ingress-nginx/tree/master/cmd/nginx) directory

> ### Version

Reviewer: ```suggestion
### Version

```

Is the package of the code responsible for adding `version` subcommand, and can be found in [version](https://github.com/kubernetes/ingress-nginx/tree/master/version) directory

### Internal code

This part of the code contains the internal logics that compose Ingress NGINX Controller, and it's split in:

#### Admission Controller

Contains the code of [Kubernetes Admission Controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) which validates the syntax of ingress objects before accepting it, and mutates (TODO: confirm if it mutates) the object if necessary.

This code can be found in [internal/admission/controller](https://github.com/kubernetes/ingress-nginx/tree/master/internal/admission/controller) directory

#### File functions

Contains auxiliary codes that deal with files, such as generating the SHA1 checksum of a file, or creating required directories.

This code can be found in [internal/file](https://github.com/kubernetes/ingress-nginx/blob/master/internal/file) directory

> #### Ingress functions

Reviewer: ```suggestion
#### Ingress functions

```

Contains all the logics from NGINX Ingress Controller, with some examples being:

* Expected Golang structures that will be used in templates and other parts of the codes - [internal/ingress/types.go](https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/types.go)

> * supported annotations and it's parsing logics - [internal/ingress/annotations](https://github.com/kubernetes/ingress-nginx/tree/master/internal/ingress/annotations)

Reviewer: ```suggestion
* supported annotations and its parsing logics - [internal/ingress/annotations](https://github.com/kubernetes/ingress-nginx/tree/master/internal/ingress/annotations)
```

This part of the code can be found in [internal/k8s](https://github.com/kubernetes/ingress-nginx/tree/master/internal/k8s) directory.

#### Networking functions

Contains helper functions for networking, such as IPv4 and IPv6 parsing, SSL certificate parsing, etc.

This part of the code can be found in [internal/net](https://github.com/kubernetes/ingress-nginx/tree/master/internal/net) directory.

#### NGINX functions

Contains helper function to deal with NGINX, such as verify if it's running and reading it's configuration file parts.

This part of the code can be found in [internal/nginx](https://github.com/kubernetes/ingress-nginx/tree/master/internal/nginx) directory.

#### Tasks / Queue

Contains the functions responsible for the sync queue part of the controller.

This part of the code can be found in [internal/task](https://github.com/kubernetes/ingress-nginx/tree/master/internal/task) directory.

#### Other parts of internal

Other parts of internal code might not be covered here, like runtime and watch but they can be added in a future.

## E2E Test

The e2e tests code is in [test](https://github.com/kubernetes/ingress-nginx/tree/master/test) directory.

> ## Other programs

Reviewer: ```suggestion
## Other programs

```

Describe here `kubectl plugin`, `dbg`, `waitshutdown` and cover the hack scripts

## Deploy files

This directory contains the `yaml` deploy files used as examples or references in the docs to deploy Ingress NGINX and other componentes.

Those files are in [deploy](https://github.com/kubernetes/ingress-nginx/tree/master/deploy) directory.

> ## Helm Chart

Reviewer: ```suggestion
## Helm Chart

```

## Documentation/Website

The documentation used to generate the website https://kubernetes.github.io/ingress-nginx/

This code is available in [docs](https://github.com/kubernetes/ingress-nginx/tree/master/docs) and it's main "language" is `Markdown`, used by [mkdocs](https://github.com/kubernetes/ingress-nginx/blob/master/mkdocs.yml) file to generate static pages

## Container Images

Container images used to run ingress-nginx, or to build the final image.

### Base Images

Contains the `Dockerfiles` and scripts used to build base images that are used in other parts of the repo. They are present in [images](https://github.com/kubernetes/ingress-nginx/tree/master/images) repo. Some examples:
* [nginx](https://github.com/kubernetes/ingress-nginx/tree/master/images/nginx) - The base NGINX image ingress-nginx uses is not a vanilla NGINX. It bundles many libraries together and it is a job in itself to maintain that and keep things up-to-date.
* [custom-error-pages](https://github.com/kubernetes/ingress-nginx/tree/master/images/custom-error-pages) - Used on the custom error page examples

There are other images inside this directory

> ### Ingress Controller Image

Reviewer: ```suggestion
### Ingress Controller Image

```

The image used to build the final ingress controller, used in deploy scripts and Helm charts. 

This is NGINX with some Lua enhancement. We do dynamic certificate, endpoints handling, canary traffic split, custom load balancing etc at this component. One can also add new functionalities using Lua plugin system.

The files are in [rootfs](https://github.com/kubernetes/ingress-nginx/tree/master/rootfs) directory and contains:

* The Dockerfile
* [Auxiliary scripts](https://github.com/kubernetes/ingress-nginx/tree/master/rootfs/ingress-controller)

> #### Ingress NGINX Lua Scripts

Reviewer: ```suggestion
#### Ingress NGINX Lua Scripts

```

> Ingress NGINX uses Lua Scripts to enable features like hot reloads, rate limits and monitoring. Some are written using the [OpenResty](https://openresty.org/en/) helper.

Reviewer: ```suggestion
Ingress NGINX uses Lua Scripts to enable features like hot reloading, rate limiting and monitoring. Some are written using the [OpenResty](https://openresty.org/en/) helper.
```

The directory containing Lua scripts is [rootfs/etc/nginx/lua](https://github.com/kubernetes/ingress-nginx/tree/master/rootfs/etc/nginx/lua)

> #### Nginx Go template file

Reviewer: ```suggestion
#### Nginx Go template file

```
